esti/copy_test.go

```diff
@@ -0,0 +1,195 @@
+package esti
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/go-test/deep"
+	"github.com/spf13/viper"
+	"github.com/stretchr/testify/require"
+	"github.com/treeverse/lakefs/pkg/api"
+	"github.com/treeverse/lakefs/pkg/block"
+	"github.com/treeverse/lakefs/pkg/block/azure"
+	"github.com/treeverse/lakefs/pkg/config"
+)
+
+const (
+	s3CopyDataPath            = "s3://esti-system-testing-data/copy-test-data/"
+	gsCopyDataPath            = "gs://esti-system-testing-data/copy-test-data/"
+	azureCopyDataPathTemplate = "https://%s.blob.core.windows.net/esti-system-testing-data/copy-test-data/"
+	azureAbortAccount         = "esti4multipleaccounts"
+	ingestionBranch           = "test-data"
+	largeObject               = "squash.tar"
+)
+
+var azureCopyDataPath = ""
+
+func TestCopyObject(t *testing.T) {
+	ctx, _, repo := setupTest(t)
+	defer tearDownTest(repo)
+
+	t.Run("copy_large_size_file", func(t *testing.T) {
+		blockstoreType := viper.GetString(config.BlockstoreTypeKey)
+		// Copying from same account occurs immediately even for large files (async)
+		if blockstoreType == block.BlockstoreTypeAzure { // Extract storage account
+			resp, err := client.GetRepositoryWithResponse(ctx, repo)
+			require.NoError(t, err)
+			require.Equal(t, http.StatusOK, resp.StatusCode())
+			u, err := url.Parse(resp.JSON200.StorageNamespace)
+			require.NoError(t, err)
+			accountName, err := azure.ExtractStorageAccount(u)
+			require.NoError(t, err)
+			azureCopyDataPath = fmt.Sprintf(azureCopyDataPathTemplate, accountName)
+		}
+
+		importTestData(t, ctx, client, repo)
+		res, err := client.StatObjectWithResponse(ctx, repo, ingestionBranch, &api.StatObjectParams{
+			Path: largeObject,
+		})
+		require.NoError(t, err)
+		require.Equal(t, http.StatusOK, res.StatusCode())
+
+		objStat := res.JSON200
+		destPath := "foo"
+		srcBranch := ingestionBranch
+		copyResp, err := client.CopyObjectWithResponse(ctx, repo, "main", &api.CopyObjectParams{
+			DestPath: destPath,
+		}, api.CopyObjectJSONRequestBody{
+			SrcPath: largeObject,
+			SrcRef:  &srcBranch,
+		})
+		require.NoError(t, err, "failed to copy")
+		require.Equal(t, http.StatusCreated, copyResp.StatusCode())
```

Comment:
Check copyResp.JSON201 is not nil - it will verify the status code and content type of the response.